README
NAME
    SOAP::Lite - Perl's Web Services Toolkit

DESCRIPTION
    HEY THERE - THIS REPO HAS MOVED HERE - https://github.com/redhotpenguin/perl-soaplite

    SOAP::Lite is a collection of Perl modules which provides a simple and
    lightweight interface to the Simple Object Access Protocol (SOAP) both
    on client and server side.

PERL VERSION WARNING
    As of version SOAP::Lite version 1.0, no perl versions before 5.8 will
    be supported.

    SOAP::Lite 0.71 will be the last version of SOAP::Lite running on perl
    5.005

    Future versions of SOAP::Lite will require at least perl 5.6.0

    If you have not had the time to upgrade your perl, you should consider
    this now.

OVERVIEW OF CLASSES AND PACKAGES
    lib/SOAP/Lite.pm
        SOAP::Lite - Main class provides all logic

        SOAP::Transport - Transport backend

        SOAP::Data - Data objects

        SOAP::Header - Header Data Objects

        SOAP::Serializer - Serializes data structures to SOAP messages

        SOAP::Deserializer - Deserializes SOAP messages into SOAP::SOM
        objects

        SOAP::SOM - SOAP Message objects

        SOAP::Constants - Provides access to common constants and defaults

        SOAP::Trace - Tracing facilities

        SOAP::Schema - Provides access and stub(s) for schema(s)

        SOAP::Schema::WSDL - WSDL implementation for SOAP::Schema

        SOAP::Server - Handles requests on server side

        SOAP::Server::Object - Handles objects-by-reference

        SOAP::Fault - Provides support for Faults on server side

        SOAP::Utils - A set of private and public utility subroutines

    lib/SOAP/Packager.pm
        SOAP::Packager - Provides an abstract class for implementing custom
        packagers.

        SOAP::Packager::MIME - Provides MIME support to SOAP::Lite

        SOAP::Packager::DIME - Provides DIME support to SOAP::Lite

    lib/SOAP/Transport/HTTP.pm
        SOAP::Transport::HTTP::Client - Client interface to HTTP transport

        SOAP::Transport::HTTP::Server - Server interface to HTTP transport

        SOAP::Transport::HTTP::CGI - CGI implementation of server interface

        SOAP::Transport::HTTP::Daemon - Daemon implementation of server
        interface

        SOAP::Transport::HTTP::Apache - mod_perl implementation of server
        interface

    lib/SOAP/Transport/POP3.pm
        SOAP::Transport::POP3::Server - Server interface to POP3 protocol

    lib/SOAP/Transport/MAILTO.pm
        SOAP::Transport::MAILTO::Client - Client interface to SMTP/sendmail

    lib/SOAP/Transport/LOCAL.pm
        SOAP::Transport::LOCAL::Client - Client interface to local transport

    lib/SOAP/Transport/TCP.pm
        SOAP::Transport::TCP::Server - Server interface to TCP protocol

        SOAP::Transport::TCP::Client - Client interface to TCP protocol

    lib/SOAP/Transport/IO.pm
        SOAP::Transport::IO::Server - Server interface to IO transport

METHODS
    All accessor methods return the current value when called with no
    arguments, while returning the object reference itself when called with
    a new value. This allows the set-attribute calls to be chained together.

    new(optional key/value pairs)
            $client = SOAP::Lite->new(proxy => $endpoint)

        Constructor. Many of the accessor methods defined here may be
        initialized at creation by providing their name as a key, followed
        by the desired value. The example provides the value for the proxy
        element of the client.

    transport(optional transport object)
            $transp = $client->transport( );

        Gets or sets the transport object used for sending/receiving SOAP
        messages.

        See SOAP::Transport for details.

    serializer(optional serializer object)
            $serial = $client->serializer( )

        Gets or sets the serializer object used for creating XML messages.

        See SOAP::Serializer for details.

    packager(optional packager object)
            $packager = $client->packager( )

        Provides access to the "SOAP::Packager" object that the client uses
        to manage the use of attachments. The default packager is a MIME
        packager, but unless you specify parts to send, no MIME formatting
        will be done.

        See also: SOAP::Packager.

    proxy(endpoint, optional extra arguments)
            $client->proxy('http://soap.xml.info/ endPoint');

        The proxy is the server or endpoint to which the client is going to
        connect. This method allows the setting of the endpoint, along with
        any extra information that the transport object may need when
        communicating the request.

        This method is actually an alias to the proxy method of
        SOAP::Transport. It is the same as typing:

            $client->transport( )->proxy(...arguments);

        Extra parameters can be passed to proxy() - see below.

        compress_threshold
            See COMPRESSION in HTTP::Transport.

        All initialization options from the underlying transport layer
            The options for HTTP(S) are the same as for LWP::UserAgent's
            new() method.

            A common option is to create a instance of HTTP::Cookies and
            pass it as cookie_jar option:

             my $cookie_jar = HTTP::Cookies->new()
             $client->proxy('http://www.example.org/webservice',
                cookie_jar => $cookie_jar,
             );

        For example, if you wish to set the HTTP timeout for a SOAP::Lite
        client to 5 seconds, use the following code:

          my $soap = SOAP::Lite
           ->uri($uri)
           ->proxy($proxyUrl, timeout => 5 );

        See LWP::UserAgent.

    endpoint(optional new endpoint address)
            $client->endpoint('http://soap.xml.info/ newPoint')

        It may be preferable to set a new endpoint without the additional
        work of examining the new address for protocol information and
        checking to ensure the support code is loaded and available. This
        method allows the caller to change the endpoint that the client is
        currently set to connect to, without reloading the relevant
        transport code. Note that the proxy method must have been called
        before this method is used.

    service(service URL)
            $client->service('http://svc.perl.org/Svc.wsdl');

        "SOAP::Lite" offers some support for creating method stubs from
        service descriptions. At present, only WSDL support is in place.
        This method loads the specified WSDL schema and uses it as the basis
        for generating stubs.

    outputxml(boolean)
            $client->outputxml('true');

        When set to a true value, the raw XML is returned by the call to a
        remote method.

        The default is to return the a SOAP::SOM object (false).

    autotype(boolean)
            $client->autotype(0);

        This method is a shortcut for:

            $client->serializer->autotype(boolean);

        By default, the serializer tries to automatically deduce types for
        the data being sent in a message. Setting a false value with this
        method disables the behavior.

    readable(boolean)
            $client->readable(1);

        This method is a shortcut for:

            $client->serializer->readable(boolean);

        When this is used to set a true value for this property, the
        generated XML sent to the endpoint has extra characters (spaces and
        new lines) added in to make the XML itself more readable to human
        eyes (presumably for debugging). The default is to not send any
        additional characters.

    default_ns($uri)
        Sets the default namespace for the request to the specified uri.
        This overrides any previous namespace declaration that may have been
        set using a previous call to "ns()" or "default_ns()". Setting the
        default namespace causes elements to be serialized without a
        namespace prefix, like this:

          <soap:Envelope>
            <soap:Body>
              <myMethod xmlns="http://www.someuri.com">
                <foo />
              </myMethod>
            </soap:Body>
          </soap:Envelope>

        Some .NET web services have been reported to require this XML
        namespace idiom.

    ns($uri,$prefix=undef)
        Sets the namespace uri and optionally the namespace prefix for the
        request to the specified values. This overrides any previous
        namespace declaration that may have been set using a previous call
        to "ns()" or "default_ns()".

        If a prefix is not specified, one will be generated for you
        automatically. Setting the namespace causes elements to be
        serialized with a declared namespace prefix, like this:

          <soap:Envelope>
            <soap:Body>
              <my:myMethod xmlns:my="http://www.someuri.com">
                <my:foo />
              </my:myMethod>
            </soap:Body>
          </soap:Envelope>

    use_prefix(boolean)
        Deprecated. Use the "ns()" and "default_ns" methods described above.

        Shortcut for "serializer->use_prefix()". This lets you turn on/off
        the use of a namespace prefix for the children of the /Envelope/Body
        element. Default is 'true'.

        When use_prefix is set to 'true', serialized XML will look like
        this:

          <SOAP-ENV:Envelope ...attributes skipped>
            <SOAP-ENV:Body>
              <namesp1:mymethod xmlns:namesp1="urn:MyURI" />
            </SOAP-ENV:Body>
          </SOAP-ENV:Envelope>

        When use_prefix is set to 'false', serialized XML will look like
        this:

          <SOAP-ENV:Envelope ...attributes skipped>
            <SOAP-ENV:Body>
              <mymethod xmlns="urn:MyURI" />
            </SOAP-ENV:Body>
          </SOAP-ENV:Envelope>

        Some .NET web services have been reported to require this XML
        namespace idiom.

    soapversion(optional value)
            $client->soapversion('1.2');

        If no parameter is given, returns the current version of SOAP that
        is being used by the client object to encode requests. If a
        parameter is given, the method attempts to set that as the version
        of SOAP being used.

        The value should be either 1.1 or 1.2.

    envprefix(QName)
            $client->envprefix('env');

        This method is a shortcut for:

            $client->serializer->envprefix(QName);

        Gets or sets the namespace prefix for the SOAP namespace. The
        default is SOAP.

        The prefix itself has no meaning, but applications may wish to chose
        one explicitly to denote different versions of SOAP or the like.

    encprefix(QName)
            $client->encprefix('enc');

        This method is a shortcut for:

            $client->serializer->encprefix(QName);

        Gets or sets the namespace prefix for the encoding rules namespace.
        The default value is SOAP-ENC.

    While it may seem to be an unnecessary operation to set a value that
    isn't relevant to the message, such as the namespace labels for the
    envelope and encoding URNs, the ability to set these labels explicitly
    can prove to be a great aid in distinguishing and debugging messages on
    the server side of operations.

    encoding(encoding URN)
            $client->encoding($soap_12_encoding_URN);

        This method is a shortcut for:

            $client->serializer->encoding(args);

        Where the earlier method dealt with the label used for the
        attributes related to the SOAP encoding scheme, this method actually
        sets the URN to be specified as the encoding scheme for the message.
        The default is to specify the encoding for SOAP 1.1, so this is
        handy for applications that need to encode according to SOAP 1.2
        rules.

    typelookup
            $client->typelookup;

        This method is a shortcut for:

            $client->serializer->typelookup;

        Gives the application access to the type-lookup table from the
        serializer object. See the section on SOAP::Serializer.

    uri(service specifier)
        Deprecated - the "uri" subroutine is deprecated in order to provide
        a more intuitive naming scheme for subroutines that set namespaces.
        In the future, you will be required to use either the "ns()" or
        "default_ns()" subroutines instead of "uri()".

            $client->uri($service_uri);

        This method is a shortcut for:

            $client->serializer->uri(service);

        The URI associated with this accessor on a client object is the
        service-specifier for the request, often encoded for HTTP-based
        requests as the SOAPAction header. While the names may seem
        confusing, this method doesn't specify the endpoint itself. In most
        circumstances, the "uri" refers to the namespace used for the
        request.

        Often times, the value may look like a valid URL. Despite this, it
        doesn't have to point to an existing resource (and often doesn't).
        This method sets and retrieves this value from the object. Note that
        no transport code is triggered by this because it has no direct
        effect on the transport of the object.

    multirefinplace(boolean)
            $client->multirefinplace(1);

        This method is a shortcut for:

            $client->serializer->multirefinplace(boolean);

        Controls how the serializer handles values that have multiple
        references to them. Recall from previous SOAP chapters that a value
        may be tagged with an identifier, then referred to in several
        places. When this is the case for a value, the serializer defaults
        to putting the data element towards the top of the message, right
        after the opening tag of the method-specification. It is serialized
        as a standalone entity with an ID that is then referenced at the
        relevant places later on. If this method is used to set a true
        value, the behavior is different. When the multirefinplace attribute
        is true, the data is serialized at the first place that references
        it, rather than as a separate element higher up in the body. This is
        more compact but may be harder to read or trace in a debugging
        environment.

    parts( ARRAY )
        Used to specify an array of MIME::Entity's to be attached to the
        transmitted SOAP message. Attachments that are returned in a
        response can be accessed by "SOAP::SOM::parts()".

    self
            $ref = SOAP::Lite->self;

        Returns an object reference to the default global object the
        "SOAP::Lite" package maintains. This is the object that processes
        many of the arguments when provided on the use line.

    The following method isn't an accessor style of method but neither does
    it fit with the group that immediately follows it:

    call(arguments)
            $client->call($method => @arguments);

        As has been illustrated in previous chapters, the "SOAP::Lite"
        client objects can manage remote calls with auto-dispatching using
        some of Perl's more elaborate features. call is used when the
        application wants a greater degree of control over the details of
        the call itself. The method may be built up from a SOAP::Data
        object, so as to allow full control over the namespace associated
        with the tag, as well as other attributes like encoding. This is
        also important for calling methods that contain characters not
        allowable in Perl function names, such as A.B.C.

    The next four methods used in the "SOAP::Lite" class are geared towards
    handling the types of events than can occur during the message
    lifecycle. Each of these sets up a callback for the event in question:

    on_action(callback)
            $client->on_action(sub { qq("$_[0]") });

        Triggered when the transport object sets up the SOAPAction header
        for an HTTP-based call. The default is to set the header to the
        string, uri#method, in which URI is the value set by the uri method
        described earlier, and method is the name of the method being
        called. When called, the routine referenced (or the closure, if
        specified as in the example) is given two arguments, uri and method,
        in that order.

        .NET web services usually expect "/" as separator for "uri" and
        "method". To change SOAP::Lite's behaviour to use uri/method as
        SOAPAction header, use the following code:

            $client->on_action( sub { join '/', @_ } );
        =item on_fault(callback)

            $client->on_fault(sub { popup_dialog($_[1]) });

        Triggered when a method call results in a fault response from the
        server. When it is called, the argument list is first the client
        object itself, followed by the object that encapsulates the fault.
        In the example, the fault object is passed (without the client
        object) to a hypothetical GUI function that presents an error dialog
        with the text of fault extracted from the object (which is covered
        shortly under the SOAP::SOM methods).

    on_nonserialized(callback)
            $client->on_nonserialized(sub { die "$_[0]?!?" });

        Occasionally, the serializer may be given data it can't turn into
        SOAP-savvy XML; for example, if a program bug results in a code
        reference or something similar being passed in as a parameter to
        method call. When that happens, this callback is activated, with one
        argument. That argument is the data item that could not be
        understood. It will be the only argument. If the routine returns,
        the return value is pasted into the message as the serialization.
        Generally, an error is in order, and this callback allows for
        control over signaling that error.

    on_debug(callback)
            $client->on_debug(sub { print @_ });

        Deprecated. Use the global +debug and +trace facilities described in
        SOAP::Trace

        Note that this method will not work as expected: Instead of
        affecting the debugging behaviour of the object called on, it will
        globally affect the debugging behaviour for all objects of that
        class.

WRITING A SOAP CLIENT
    This chapter guides you to writing a SOAP client by example.

    The SOAP service to be accessed is a simple variation of the well-known
    hello world program. It accepts two parameters, a name and a given name,
    and returns "Hello $given_name $name".

    We will use "Martin Kutter" as the name for the call, so all variants
    will print the following message on success:

     Hello Martin Kutter!

  SOAP message styles
    There are three common (and one less common) variants of SOAP messages.

    These address the message style (positional parameters vs. specified
    message documents) and encoding (as-is vs. typed).

    The different message styles are:

    *   rpc/encoded

        Typed, positional parameters. Widely used in scripting languages.
        The type of the arguments is included in the message. Arrays and the
        like may be encoded using SOAP encoding rules (or others).

    *   rpc/literal

        As-is, positional parameters. The type of arguments is defined by
        some pre-exchanged interface definition.

    *   document/encoded

        Specified message with typed elements. Rarely used.

    *   document/literal

        Specified message with as-is elements. The message specification and
        element types are defined by some pre-exchanged interface
        definition.

    As of 2008, document/literal has become the predominant SOAP message
    variant. rpc/literal and rpc/encoded are still in use, mainly with
    scripting languages, while document/encoded is hardly used at all.

    You will see clients for the rpc/encoded and document/literal SOAP
    variants in this section.

  Example implementations
   RPC/ENCODED
    Rpc/encoded is most popular with scripting languages like perl, php and
    python without the use of a WSDL. Usual method descriptions look like
    this:

     Method: sayHello(string, string)
     Parameters:
        name: string
        givenName: string

    Such a description usually means that you can call a method named
    "sayHello" with two positional parameters, "name" and "givenName", which
    both are strings.

    The message corresponding to this description looks somewhat like this:

     <sayHello xmlns="urn:HelloWorld">
       <s-gensym01 xsi:type="xsd:string">Kutter</s-gensym01>
       <s-gensym02 xsi:type="xsd:string">Martin</s-gensym02>
     </sayHello>

    Any XML tag names may be used instead of the "s-gensym01" stuff -
    parameters are positional, the tag names have no meaning.

    A client producing such a call is implemented like this:

     use SOAP::Lite;
     my $soap = SOAP::Lite->new( proxy => 'http://localhost:81/soap-wsdl-test/helloworld.pl');
     $soap->default_ns('urn:HelloWorld');
     my $som = $soap->call('sayHello', 'Kutter', 'Martin');

Changes
#
# Revision history for Perl extension SOAP::Lite.
#

The contents of this file provides a change history for the SOAP::Lite
Perl module. New features are designated with a '+' and bug fixes with
a '!'.
Numbers in square brackets are sourceforge tracker IDs from
https://sourceforge.net/tracker/?group_id=66000&atid=513017

Numbers with # are CPANRT tracker IDs from
http://rt.cpan.org/Public/Dist/Display.html?Name=SOAP-Lite

-----------------------------------------------------------------------
THIS RELEASE
-----------------------------------------------------------------------

1.26 May 14, 2018
    ! Problem with compression in server mode

1.25 December 29, 2017
    ! Bad usage of UNIVERSAL::isa (See RT#62667)

1.24 December 19, 2017
    + Consistent Versioning https://github.com/redhotpenguin/soaplite/pull/24
    ! Remove is_utf8() check from 78588

1.23 December 18, 2017
    ! RT #78588 unicode data not correct encoded

1.22 August 15, 2017
    ! Noindex URI::tcp module [github/JaSei]
    ! Correct the soap namespace declaration when calling services defined in WSDL [github/waaeer]
    ! Fix typos [github/kmhl]

1.21 November 6, 2016
    ! #118654 consistently use '#!/usr/bin/env perl' throughout (@mototimo) [github/moto-timo]

1.20 June 9, 2016
    ! Minor kwalitee updates (@oeuftete)
    ! #106688 encode utf8 content in test which fails under RHES 6.6

1.19 August 26, 2015
    ! #106665 Correct the version manually, no autoversioning, and the maintainer screwed up

1.18 August 25, 2015
    ! Fix a couple of cases where undefined values were being passed to functions [github/redhotpenguin]
    ! Do not set content_length for compressed request [github/bor]
    ! Add test for check return of send_receive(), add titles for some tests [github/bor]

1.17 July 30, 2015
    ! #106051 Fix utf8 encoding for CGI test (thanks cpan/ETHER for tips)

1.16 July 23, 2015
    ! Fix empty release. Note to self - do not release code when wife is in early labor

1.15 July 21, 2015
    + Preventing "billion laughs" exponential entity expansion attack [github/thilo-fester]

1.14 March 24, 2015
    + Add header and body tags [@bubi6608]
    ! Fixed typo in trace logging example [@oschwald]
    + Converted the dist from using ExtUtils::MakeMaker to using Dist::Zilla [@mvuets]
    ! Cleaned up prereqs [@mvuets]
    ! Cleaned up author and release tests [@mvuets]

1.13 December 30, 2014
    ! Fixed undefined value error when using bugzilla's xmlrpc.cgi (bugzilla's API) [dbeusee]

1.12 November 26, 2014
    ! #100557 Redefined sub in SOAP::Lite::Deserializer::XMLSchema2001

1.11 February 21, 2014
    ! #93080 bug when SOAP::Lite is trying to parse WSDL definition by https with custom settings

1.10 January 23, 2014 
    ! Fixed uninitialized warning when Context-Length is not set [github/dbeusee]
    ! avoid string eval in as_base64 [github/otrosien]
    ! #88915 warnings issued with using importing with +trace

1.09 January 14, 2013
    ! #92025 tighten multipart content-type matching [dmn@debian.org]

1.08 November 8, 2013
    ! #15703 add no_index for examples to prevent CPAN pollution [githumb.com/sergeyromanov]

1.07 November 7, 2013
    ! #89948 Remove warning when Content-Id is not present

1.06 August 21, 2013
    ! #87935 skip all MIME-Parser tests if it is not installed
    ! #87787 Use of uninitialized value $ns in Lite.pm 

1.05 August 18, 2013
    ! #87848 Fix manifest, fix invalid http version from release blunder

1.04 August 9, 2013
    ! #87605 Skip XML::Parser::Lite tests if it is missing [github.com/omega] 

1.03 August 4, 2013
    ! #87195 Remove DIME::Tools dep until we can work around its issues

1.02 July 30, 2013
    ! #87421 1.01 release blew out xml versions. mea culpa.

1.01 July 29, 2013
    ! #84168 Test t/02-payload.t fails under Perl 5.17.10
    ! #87127 TEST.pl removed from MANIFEST
    ! revert #85713 - various reports of environment failures

1.0 July 16, 2013
    ! #85713 SOAP::Transport::HTTP, 500 error skips parsing of response
    ! No more Windows 98 client support. Wait, wut?
    ! Merge SOAP::Transport::TCP back into SOAP::Lite (for now)
    ! #82416: xop href cid: URI encoded support reported by Michael R. Davis (mrdvt92)
    ! #85208 bad test plan
    ! #83750 different headers for mod_perl vs mod_perl2 [Mithun Bhattacharya]

0.716 May 10, 2013
    ! #17275 Client unable to detect fatal SOAP failure (e.g. unable to connect to host)
    ! #81471 Force array context for correct Apache header parsing in SOAP::Transport::HTTP
    ! #45611 deserializing SOAP message with composite attachment raises exception
    ! #84168 Test t/02-payload.t fails under Perl 5.17.10
    ! #85098 Monkey patch for LWP::Protocol confuses the toolchain
    ! #78692 / #84724 / #83715 Sending large object (many levels deep) throws error Incorrect parameter
    ! #78502 t/08-schema.t noisy under 5.17.2 (unescaped left brace in regex  is deprecated)
    ! #78608 Documentation for "+trace"
    ! #78639 Spelling error reported by Debian tools
0.715
    ! #75938 Change to warning / add configuration variable to allow XML tags starting with /xml/i
    ! #76902 Missing info in "SSL CERTIFICATE AUTHENTICATION" documentation
	! #74577 work around/fixes for success with MS WS https/compressed
    ! #74308 SOAP-Lite fails for operations with period in name
    ! #73802 client did not set the encoding of the soap-envelope to 'iso-8859-1' in HTTP-request
    ! #75019 HTTP server doesn't process PIPE signal
    ! #72836 Processing a SOAP 1.2 request breaks the next SOAP 1.1 response
    ! #70381 SOAP-Lite performance
    ! #53375 comments after <?xml .. >
    ! #63693 Uninitalized value (again)
    ! #37360 Use of uninitialized variables in SOAP tests
    ! fixed encoding in xmlrpc keys
    ! #15176 problems with SOAP::Header ->value 
0.714
    Moved SOAP::Transport::TCP into SOAP-Transport-TPC distribution
    ! #70329 - not generating attributes properly (deleting id attribute)
0.713

Makefile.PL
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

use 5.006; #weak references require perl 5.6
use strict;
use ExtUtils::MakeMaker;
use diagnostics;

my @PREREQS = (
#   ["<feature>","<installed module>",<dependency hash>,<install by default, 2=always>];
    [
        "Core Package",
        "SOAP::Lite",
        {
          "Test::More"      => 0,
          "XML::Parser"     => "2.23",
          "MIME::Base64"    => 0,
          "URI"             => 0,
          "Scalar::Util"    => 0,
          "Task::Weaken"    => 0, #to be sure than weaken works (Scalar::Util has XS part)
          "constant"        => 0, #don't force people to upgrade
          "Class::Inspector" => 0,
        },
        2
    ],
    [
        "Client HTTP support",
        "SOAP::Transport::HTTP",
        {
            "LWP::UserAgent" => 0
        },
        2
    ],
    ["Client HTTPS support","SOAP::Transport::HTTPS::Client",{"Crypt::SSLeay" => 0},0],
    ["Client SMTP/sendmail support","SOAP::Transport::MAILTO::Client",{"MIME::Lite" => 0},1],
    ["Client FTP support","SOAP::Transport::FTP::Client",{"SOAP::Transport::FTP" => 0.711},0],
    ["Client TCP support","SOAP::Transport::TCP::Client",{"SOAP::Transport::TCP" => 0.714},0],    
    ["Standalone HTTP server","SOAP::Transport::HTTP::Daemon",{"HTTP::Daemon" => 0},1],
    ["Apache/mod_perl server","SOAP::Transport::HTTP::Apache",{"Apache" => 0},0],
    ["FastCGI server","SOAP::Transport::HTTP::FCGI",{"FCGI" => 0},0],
    ["POP3 server","SOAP::Transport::POP3::Server",{"Net::POP3" => 0,"MIME::Parser" => 0},1],
    ["IO server","SOAP::Transport::IO::Server",{"IO::File" => 0},0],
    ["MQ transport support","SOAP::Transport::MQ",{"SOAP::Transport::MQ" => 0.712},0],
    ["JABBER transport support","SOAP::Transport::JABBER",{"SOAP::Transport::JABBER" => 0.712},0],
    ["MIME messages","SOAP::Packager::MIME",{"MIME::Parser" => 0},1],
    ["DIME messages","SOAP::Packager::DIME",{"IO::Scalar" => "2.105", "DIME::Tools" => 0.03, "Data::UUID" => "0.11"},0],
    ["SSL Support for TCP Transport","SOAP::Transport::TCP",{"IO::Socket::SSL" => 0},0],
    ["Compression support for HTTP","SOAP::Transport::HTTP",{"Compress::Zlib" => 0},0],
    ["MIME interoperability w/ Axis","SOAP::Lite",{"MIME::Parser" => "6.106"},0],
);

use Getopt::Long;

my $helptext = <<EOI;
Usage: perl Makefile.PL <options>

Possible options are:

  --noprompt            Disable interactive dialog
  --help, -?            Display this help text

  [Do not] install prerequisites for appropriate module:

EOI

# Create config parameters using module names and expand help text
# We will create a hash (%config) that has each module => (1|0) for install
my(%options, %config, %has_module_cache);

# Initialize the prereq table and all help text
foreach my $prereq (@PREREQS) {
    my ($feature,$dep,$modules,$default) = @$prereq;
    next unless $dep ne "";
    $prereq->[3] = has_all_modules($modules) unless $prereq->[3] == 2;

    my $module = do { $dep =~ s/::/-/g; $dep };
    my $shortcut = do { (my $t = $module) =~ s/SOAP-(?:Transport-)?//; $t };
    $config{$dep} = has_all_modules($modules);
    $options{"install-$dep|$shortcut!"} = \$config{$dep};
    $helptext .= sprintf "  --[no]install-%-28s --[no]%s\n", $dep, $shortcut;
}

use vars qw($noprompt $alltests $help $intro);
#my ($noprompt,$alltests,$help);

GetOptions (
    "noprompt" => \$noprompt,
    "help|?" => \$help,
);

# don't prompt when run on a smoke test system.
# See http://cpantest.grango.org/wiki/CPANAuthorNotes
# Thanks to Andreas J Koenig && all CPAN testers !
$noprompt++ if $ENV{AUTOMATED_TESTING};

$help and print($helptext), exit;

$intro = <<EOI;
We are about to install SOAP::Lite and for your convenience will provide
you with list of modules and prerequisites, so you'll be able to choose
only modules you need for your configuration.

XMLRPC::Lite, UDDI::Lite, and XML::Parser::Lite are included by default.
Installed transports can be used for both SOAP::Lite and XMLRPC::Lite.

EOI

if ($noprompt) {
    print "These are the modules that will get installed:\n\n";
} else {
    ExtUtils::MakeMaker::prompt($intro . "Press <enter> to see the detailed list.");
}

# This hash will contain a list of all perl modules we would like to
# explicitly depend upon in our Makefile
my %prereqs;
my $proceed = 0;
do {
    print "\n".generate_prereq_table()."\n";
    # Ask the user if this is the configuration they desire
    if ($noprompt) {
        $proceed = 1;
    } else {
        $proceed = ExtUtils::MakeMaker::prompt("Do you want to proceed with this configuration?" => 'yes') =~ /^\s*y/i;
    }
    # Loop through each prerequisite and ask the user if they wish to
    # install it or not - reset prereqs, cause they change each time
    %prereqs = ();
    for my $prereq (@PREREQS) {
        my ($feature,$dep,$modules,$default) = @$prereq;
        next unless $dep ne "";
        unless ($proceed || $prereq->[3] == 2) { # need to use $prereq, because we modify actual value
            $prereq->[3] = (ExtUtils::MakeMaker::prompt("Do you plan to use ${feature}?" => ($prereq->[3] ? 'yes' : 'no')) =~ /^\s*(y)/i);
        }
        (%prereqs = (%prereqs, map { $_,$modules->{$_} } keys %$modules)) if $prereq->[3];
    }
} while (!$proceed);


# Old way of selecting tests - replaced by extended test suite in
# subdirectories...
# my $tests = join ' ', glob ($noncoretests ? 't/*.t' : 't/0*.t');

# Additional tests require Test::More
# only run noncoretests if we have Test::More
my $noncoretests;
if (eval { require Test::More; }) {
    $noncoretests++;
}

my $tests;
if ($noncoretests) {
    my @test_from = ();
    require File::Find;
    File::Find::find(sub {
            return if -d $_;
            return if $_ !~m{\.t$}x;
            return if $File::Find::dir=~m{\bCVS\b};
            return if $File::Find::dir=~m{\b\.svn\b};
            return if $_=~m{^[1-9]}x;
            push @test_from, $File::Find::name;
        }, 't/'
    );
    $tests = join ' ' , @test_from;
}
else {
    $tests = join ' ', glob ('t/0*.t t/*xmlparserlite.t t/*loopback.t');
}

ExtUtils::MakeMaker::WriteMakefile(
    'NAME'          => 'SOAP::Lite',
    'AUTHOR'        =>  'Pavel Kulchenko, Randy J Ray, Byrne Reese, Martin Kutter (MKUTTER@cpan.org)',
    'LICENSE'       => 'perl',
    'VERSION_FROM'  => 'lib/SOAP/Lite.pm',
    'ABSTRACT_FROM' => 'lib/SOAP/Lite.pm',
    'PREREQ_PM'     => {
        %prereqs
     },
    'EXE_FILES'     => ['bin/SOAPsh.pl', 'bin/XMLRPCsh.pl', 'bin/stubmaker.pl'],
    test            => {TESTS => $tests},
    ($ExtUtils::MakeMaker::VERSION ge '6.48'? 
     (MIN_PERL_VERSION => 5.006,
      META_MERGE => {
          resources=> {
              repository => 'https://github.com/redhotpenguin/soaplite',
          },
          keywords => ['SOAP','SOAP client','SOAP server'],
          
      },
     ) : ()),
);


######################################################
# Supporting subroutines
######################################################
# Maintains a cache of what 3rd party modules you have
# installed
sub has_module {
    my ($mod, $version) = @_;
    $version ||= '';
    # use "require Foo; Exporter::require_version('Foo' => 1)" instead of
    # obvious "use Foo 1". The later doesn't work with MIME::Parser
    # wonder why? --PK
    return ($has_module_cache{"$mod$version"}
        ||= eval("require $mod; Exporter::require_version('$mod', $version) if ($version); 1"));
}

# Return 1 if all modules contained in the inputed array
# are installed, 0 otherwise
sub has_all_modules {
    my ($mods) = @_;
    foreach my $mod (keys %$mods) {
        return 0 if !has_module($mod, $mods->{$mod});
    }
    return 1;
}

# Print a single prerequisite to the screen
sub generate_prereq {
    my ($feature,$dep,$modules,$install) = @_;
    my $i = 0;
    my $buf = "";
    foreach my $module (keys %$modules) {
        my $detected = (has_module($module, $modules->{$module}) ? "*" : " ");
        $buf .= sprintf("%-29s [%s] %-24s %-8s\n",($i++ ? "" : $feature),$detected,$module . ($modules->{$module} ? " (v$modules->{$module})" : ""),($i == 1 ? ($install ? ($install == 2 ? "always" : "[ yes ]") : "[ no ]") : ""));
    }
    return $buf;
}

# Print the entire prerequisites table
sub generate_prereq_table {
    my $buf = sprintf("%-29s %-28s %-8s\n","Feature","Prerequisites","Install?");
    $buf .= sprintf("%s %s %s\n","-" x 29,"-" x 28,"-" x 8);
    foreach my $pre (@PREREQS) {
        my ($feature,$dep,$modules,$default) = @{$pre};
        $buf .= generate_prereq(@{$pre});
    }
    $buf .= "--- An asterix '[*]' indicates if the module is currently installed.\n";
    return $buf;
}

1;
